code/modules/mob/living/living.dm

```diff
@@ -1011,10 +1010,29 @@
 				if((N.health + N.halloss) < config.health_threshold_crit || N.stat == DEAD)
 					N.adjustBruteLoss(rand(10,30))
 			src.drop_from_inventory(G)
+			return TRUE
+		else
+			return FALSE
+
+	if(!item)
+		return FALSE //Grab processing has a chance of returning null
+
+	if(a_intent == I_HELP && Adjacent(target) && isitem(item))
+		var/obj/item/I = item
+		if(ishuman(target))
+			var/mob/living/carbon/human/H = target
+			if(H.in_throw_mode && H.a_intent == I_HELP && unEquip(I))
+				H.put_in_hands(I) // If this fails it will just end up on the floor, but that's fitting for things like dionaea.
+				visible_message("<b>[src]</b> hands \the [H] \a [I].", SPAN_NOTICE("You give \the [target] \a [I]."))
+			else
+				to_chat(src, SPAN_NOTICE("You offer \the [I] to \the [target]."))
+				do_give(H)
+			return TRUE
+		make_item_drop_sound(I)
+		I.forceMove(get_turf(target))
+		return TRUE
 
-	src.drop_from_inventory(item)
-	if(!item || !isturf(item.loc))
-		return
+	drop_from_inventory(item)
```

Comment:
Need to check `!item || QDELETED(item)` after this for stuff like ninja sword that deletes itself on drop.